kernel: userspace: reserve stack space to store local data

dcpleung opened the pull request:
This enables reserving little space on the top of stack to store
data local to thread when CONFIG_USERSPACE. The first customer
of this is errno.

Note that ARC, due to how it lays out the user stack and
privilege stack, sets the pointer itself rather than
relying on the common way.

Fixes: #9067

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

andyross reviewed (changes requested):
Some whining.  If I'm right about the alignment thing that needs to be fixed, the other stuff is just nitpicking.

dcpleung commented on include/kernel.h:
```diff
@@ -538,14 +548,12 @@ struct k_thread {
 #endif
 
+#ifdef CONFIG_THREAD_USERSPACE_LOCAL_DATA
+	struct _thread_userspace_local_data *userspace_local_data;
```
ARC's way of placing the user stack and privilege stack disallows using the stack_size trick to reserve space for the struct. This means the ((char *)stack + stack_size) would not work on ARC, and thus needing this. AFAICT, errno is a macro to (*z_errno()) which is a syscall. Would it still always be cached?

dcpleung committed:
kernel: userspace: reserve stack space to store local data

This enables reserving little space on the top of stack to store
data local to thread when CONFIG_USERSPACE. The first customer
of this is errno.

Note that ARC, due to how it lays out the user stack and
privilege stack, sets the pointer itself rather than
relying on the common way.

Fixes: #9067

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung commented on kernel/thread.c:
```diff
@@ -294,8 +294,26 @@ void _setup_new_thread(struct k_thread *new_thread,
+	/* reserve space on top of stack for local data */
+	stack_size -= THREAD_USERSPACE_LOCAL_DATA_SIZE;
+	new_thread->userspace_local_data = NULL;
+#endif
```
The new patch has the alignment.

dcpleung commented on include/kernel.h:
```diff
@@ -503,6 +503,16 @@ struct _mem_domain_info {
+};
+
+#define THREAD_USERSPACE_LOCAL_DATA_SIZE \
+	(STACK_ROUND_UP(sizeof(struct _thread_userspace_local_data)))
```
Removed in new patch since the alignment is done elsewhere.
